src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs

```diff
@@ -184,64 +193,104 @@ private static void AdjustSpans(ArrayBuilder<ClassifiedSpan> spans, TextSpan wid
             }
         }
 
-        private static ImmutableArray<ClassifiedSpan> MergeParts(
+        private readonly struct ClassifiedSpanIntervalIntrospector : IIntervalIntrospector<ClassifiedSpan>
+        {
+            public static readonly ClassifiedSpanIntervalIntrospector Instance = new ClassifiedSpanIntervalIntrospector();
+
+            public int GetLength(ClassifiedSpan value)
+                => value.TextSpan.Length;
+
+            public int GetStart(ClassifiedSpan value)
+                => value.TextSpan.Start;
+        }
+
+        /// <summary>
+        /// Adds all semantic parts to final parts, and adds all portions of <paramref name="syntaxParts"/> that do not
+        /// overlap with any semantic parts as well.  All final parts will be non-empty.
+        /// </summary>
+        private static void MergeParts(
             ArrayBuilder<ClassifiedSpan> syntaxParts,
-            ArrayBuilder<ClassifiedSpan> semanticParts)
+            ArrayBuilder<ClassifiedSpan> semanticParts,
+            ArrayBuilder<ClassifiedSpan> finalParts)
         {
-            // Take all the syntax parts.  However, if any have been overridden by a 
-            // semantic part, then choose that one.
+            // Create an interval tree so we can easily determine which semantic parts intersect with the 
+            // syntactic parts we're looking at.
+            var semanticPartsTree = new SimpleIntervalTree<ClassifiedSpan, ClassifiedSpanIntervalIntrospector>(
+                ClassifiedSpanIntervalIntrospector.Instance, values: null);
```

Comment:
an interval tree is a nice data structure here as it gives us an easy way to find all semantic classifications that might overlap with a particular syntactic one.  this is the same data structure we use for tagging.